<style>
/* notebook-style-v1: dark-theme-first markdown presentation */

.jp-MarkdownCell .jp-RenderedHTMLCommon,
.text_cell_render,
.markdown-body {
  color: #d7e0e7;
  line-height: 1.8;
}

.jp-MarkdownCell .jp-RenderedHTMLCommon h1,
.text_cell_render h1,
.markdown-body h1 {
  background: linear-gradient(135deg, #22313d 0%, #284740 100%);
  color: #edf5f3;
  padding: 18px 22px;
  border-radius: 20px;
  border: 1px solid #3a5255;
  border-left: 10px solid #78b0a1;
  box-shadow: 0 10px 24px rgba(0, 0, 0, 0.20);
  margin: 0 0 18px 0;
}

.jp-MarkdownCell .jp-RenderedHTMLCommon h2,
.text_cell_render h2,
.markdown-body h2 {
  background: linear-gradient(90deg, #22313d 0%, #273743 100%);
  border-left: 8px solid #72aa9b;
  color: #e7eff3;
  padding: 11px 16px;
  border-radius: 14px;
  border: 1px solid #3a4b58;
  box-shadow: 0 6px 18px rgba(0, 0, 0, 0.18);
  margin: 22px 0 14px 0;
}

.jp-MarkdownCell .jp-RenderedHTMLCommon h3,
.text_cell_render h3,
.markdown-body h3 {
  display: inline-block;
  color: #d8e4ea;
  background: #22303a;
  border-left: 5px solid #7fa89e;
  padding: 8px 14px;
  border-radius: 12px;
  border: 1px solid #3a4a52;
  margin: 18px 0 10px 0;
}

.jp-MarkdownCell .jp-RenderedHTMLCommon h4,
.text_cell_render h4,
.markdown-body h4 {
  color: #9fd2c3;
  margin: 16px 0 8px 0;
  font-weight: 700;
}

.jp-MarkdownCell .jp-RenderedHTMLCommon p,
.jp-MarkdownCell .jp-RenderedHTMLCommon li,
.text_cell_render p,
.text_cell_render li,
.markdown-body p,
.markdown-body li {
  color: #d7e0e7;
  line-height: 1.8;
}

.jp-MarkdownCell .jp-RenderedHTMLCommon ul,
.jp-MarkdownCell .jp-RenderedHTMLCommon ol,
.text_cell_render ul,
.text_cell_render ol,
.markdown-body ul,
.markdown-body ol {
  margin: 10px 0 16px 20px;
  padding-left: 18px;
}

.jp-MarkdownCell .jp-RenderedHTMLCommon li,
.text_cell_render li,
.markdown-body li {
  margin: 6px 0;
}

.jp-MarkdownCell .jp-RenderedHTMLCommon blockquote,
.text_cell_render blockquote,
.markdown-body blockquote {
  margin: 14px 0;
  padding: 12px 16px;
  background: #24313d;
  border-left: 5px solid #90b7a2;
  border-radius: 14px;
  border: 1px solid #3c4d5a;
  color: #d9e4ec;
}

.jp-MarkdownCell .jp-RenderedHTMLCommon blockquote p,
.text_cell_render blockquote p,
.markdown-body blockquote p {
  margin: 6px 0;
}

.jp-MarkdownCell .jp-RenderedHTMLCommon pre,
.text_cell_render pre,
.markdown-body pre {
  background: #1b2330;
  color: #e6edf3;
  padding: 14px 16px;
  border-radius: 14px;
  border: 1px solid #334155;
  overflow-x: auto;
}

.jp-MarkdownCell .jp-RenderedHTMLCommon code,
.text_cell_render code,
.markdown-body code {
  background: #27313a;
  color: #e8eef4;
  padding: 2px 6px;
  border-radius: 6px;
  border: 1px solid #3c4a55;
}

.jp-MarkdownCell .jp-RenderedHTMLCommon pre code,
.text_cell_render pre code,
.markdown-body pre code {
  background: transparent;
  border: 0;
  padding: 0;
}

.jp-MarkdownCell .jp-RenderedHTMLCommon hr,
.text_cell_render hr,
.markdown-body hr {
  border: 0;
  height: 1px;
  background: linear-gradient(90deg, transparent, #556977, transparent);
  margin: 20px 0;
}

.jp-MarkdownCell .jp-RenderedHTMLCommon table,
.text_cell_render table,
.markdown-body table {
  border-collapse: collapse;
  background: #ffffff;
  color: #31475b;
  font-size: 0.96rem;
  margin: 16px 0;
}

.jp-MarkdownCell .jp-RenderedHTMLCommon thead th,
.text_cell_render thead th,
.markdown-body thead th {
  background: #f3f6f8;
  color: #24415c;
  padding: 10px 12px;
  text-align: left;
  border: 1px solid #dde6ec;
  font-weight: 700;
}

.jp-MarkdownCell .jp-RenderedHTMLCommon tbody td,
.text_cell_render tbody td,
.markdown-body tbody td {
  padding: 10px 12px;
  border: 1px solid #e4edf2;
  vertical-align: top;
  color: #31475b;
}

.jp-MarkdownCell .jp-RenderedHTMLCommon tbody tr:nth-child(even) td,
.text_cell_render tbody tr:nth-child(even) td,
.markdown-body tbody tr:nth-child(even) td {
  background: #f8fbff;
}

.jp-MarkdownCell .jp-RenderedHTMLCommon a,
.text_cell_render a,
.markdown-body a {
  color: #8fd1ff;
}

.jp-MarkdownCell .jp-RenderedHTMLCommon summary,
.text_cell_render summary,
.markdown-body summary {
  color: #e6edf3;
  cursor: pointer;
  font-weight: 700;
}

.jp-MarkdownCell .jp-RenderedHTMLCommon mjx-container,
.text_cell_render mjx-container,
.markdown-body mjx-container {
  color: #e6edf3 !important;
}
</style>
<div data-notebook-style-version="2026-04-v1" style="display:none;"></div>


# Pandas Cheat Sheet Notebook  
## Advanced `groupby` explained from the start

This notebook is a clean interview-prep cheat sheet using **one dataframe** and solving **Tasks 1 to 7** with:
- explanation
- code
- visible output
- interview meaning

Topics covered:
1. Department-wise salary summary  
2. Department and city wise average salary  
3. Add department average salary to each row  
4. Compare employee salary to department average  
5. Count high performers by department  
6. Percentage of high performers by department  
7. Top paid employee in each department


## 0. Setup

We start by importing pandas and numpy.
- `pandas` is for dataframe operations
- `numpy` helps with conditional columns like `np.where`


In [1]:
import pandas as pd
import numpy as np

## 1. Create the dataframe

This is the single dataframe used for all questions.

Important columns:
- `dept`: department
- `city`: location
- `salary`: fixed pay
- `bonus`: extra pay
- `rating`: performance rating
- `skills`: list column
- `projects`: nested list of dictionaries

We also convert `joined` into datetime.


In [2]:
df = pd.DataFrame({
    "emp_id": [101, 102, 103, 104, 105, 106, 107, 108, 109, 110],
    "name": ["Alice", "Bob", "Charlie", "David", "Eva", "Frank", "Grace", "Helen", "Ian", "Jane"],
    "dept": ["Data", "Data", "Web", "Web", "Data", "ML", "ML", "Web", "Data", "ML"],
    "city": ["Bangalore", "Bangalore", "Delhi", "Delhi", "Mumbai", "Bangalore", "Mumbai", "Delhi", "Mumbai", "Bangalore"],
    "salary": [120000, 95000, 110000, 105000, 130000, 150000, 140000, 99000, 125000, 155000],
    "bonus": [10000, 8000, 12000, 7000, 15000, 20000, 18000, 6000, 14000, 22000],
    "rating": [5, 3, 4, 2, 5, 4, 5, 3, 4, 5],
    "experience": [6, 3, 5, 2, 7, 8, 6, 3, 5, 9],
    "joined": ["2020-01-15", "2021-06-10", "2019-09-23", "2022-03-01", "2018-11-12",
               "2017-07-19", "2020-08-25", "2021-01-11", "2019-05-30", "2016-12-05"],
    "skills": [
        ["python", "sql", "pandas"],
        ["python", "excel"],
        ["javascript", "react"],
        ["javascript", "css"],
        ["python", "sql", "spark"],
        ["python", "ml", "deep-learning"],
        ["python", "ml", "sql"],
        ["html", "css", "javascript"],
        ["python", "sql", "tableau"],
        ["python", "ml", "nlp"]
    ],
    "projects": [
        [{"name": "fraud", "hours": 120}, {"name": "churn", "hours": 80}],
        [{"name": "sales_dashboard", "hours": 60}],
        [{"name": "website_revamp", "hours": 100}],
        [{"name": "landing_page", "hours": 40}, {"name": "seo_fix", "hours": 30}],
        [{"name": "recommendation", "hours": 150}],
        [{"name": "forecasting", "hours": 200}, {"name": "vision", "hours": 90}],
        [{"name": "classification", "hours": 170}],
        [{"name": "frontend_migration", "hours": 110}],
        [{"name": "reporting", "hours": 95}, {"name": "automation", "hours": 50}],
        [{"name": "llm_search", "hours": 220}, {"name": "chatbot", "hours": 140}]
    ]
})

df["joined"] = pd.to_datetime(df["joined"])
df

,emp_id,name,dept,city,salary,bonus,rating,experience,joined,skills,projects
0,101,Alice,Data,Bangalore,120000,10000,5,6,2020-01-15,"[python, sql, pandas]","[{'name': 'fraud', 'hours': 120}, {'name': 'ch..."
1,102,Bob,Data,Bangalore,95000,8000,3,3,2021-06-10,"[python, excel]","[{'name': 'sales_dashboard', 'hours': 60}]"
2,103,Charlie,Web,Delhi,110000,12000,4,5,2019-09-23,"[javascript, react]","[{'name': 'website_revamp', 'hours': 100}]"
3,104,David,Web,Delhi,105000,7000,2,2,2022-03-01,"[javascript, css]","[{'name': 'landing_page', 'hours': 40}, {'name..."
4,105,Eva,Data,Mumbai,130000,15000,5,7,2018-11-12,"[python, sql, spark]","[{'name': 'recommendation', 'hours': 150}]"
5,106,Frank,ML,Bangalore,150000,20000,4,8,2017-07-19,"[python, ml, deep-learning]","[{'name': 'forecasting', 'hours': 200}, {'name..."
6,107,Grace,ML,Mumbai,140000,18000,5,6,2020-08-25,"[python, ml, sql]","[{'name': 'classification', 'hours': 170}]"
7,108,Helen,Web,Delhi,99000,6000,3,3,2021-01-11,"[html, css, javascript]","[{'name': 'frontend_migration', 'hours': 110}]"
8,109,Ian,Data,Mumbai,125000,14000,4,5,2019-05-30,"[python, sql, tableau]","[{'name': 'reporting', 'hours': 95}, {'name': ..."
9,110,Jane,ML,Bangalore,155000,22000,5,9,2016-12-05,"[python, ml, nlp]","[{'name': 'llm_search', 'hours': 220}, {'name'..."


## Quick reminder: what `groupby` does

`groupby` means:
1. split rows into groups
2. apply a function to each group
3. combine the results

### Common patterns
- `agg()` -> one row per group
- `transform()` -> same number of rows as original dataframe
- `filter()` -> keep or remove whole groups
- `apply()` -> custom logic on each group


## 2. Task 1 — Department-wise salary summary

### Goal

For each `dept`, calculate:
- total salary
- average salary
- max salary
- min salary
- employee count

### Why this uses `agg`
We want **one summary row per department**, so `groupby(...).agg(...)` is the right tool.


In [3]:
dept_salary_summary = df.groupby("dept").agg(
    total_salary=("salary", "sum"),
    avg_salary=("salary", "mean"),
    max_salary=("salary", "max"),
    min_salary=("salary", "min"),
    employee_count=("emp_id", "count")
).reset_index()

dept_salary_summary

,dept,total_salary,avg_salary,max_salary,min_salary,employee_count
0,Data,470000,117500.000000,130000,95000,4
1,ML,445000,148333.333333,155000,140000,3
2,Web,314000,104666.666667,110000,99000,3


### Explanation

For `Data` department, salaries are:
- 120000
- 95000
- 130000
- 125000

So:
- total = 470000
- average = 117500
- max = 130000
- min = 95000
- count = 4

This is a classic interview pattern: **group + multiple metrics**.


## 3. Task 2 — Department and city wise average salary

### Goal

Find average salary for each combination of:
- `dept`
- `city`

### Why this uses multi-column `groupby`
Now the group is not just department.  
The group is the combination of **department + city**.


In [4]:
dept_city_avg_salary = df.groupby(["dept", "city"], as_index=False).agg(
    avg_salary=("salary", "mean")
)

dept_city_avg_salary

,dept,city,avg_salary
0,Data,Bangalore,107500.000000
1,Data,Mumbai,127500.000000
2,ML,Bangalore,152500.000000
3,ML,Mumbai,140000.000000
4,Web,Delhi,104666.666667


### Explanation

Example:
- `Data + Bangalore` has Alice and Bob
- average salary = `(120000 + 95000) / 2 = 107500`

This pattern is useful when interviewers ask:
- average by team and region
- sales by category and month
- score by class and section


## 4. Task 3 — Add department average salary to each row

### Goal

Create a new column:
- `dept_avg_salary`

### Why this uses `transform`
We want the **department average repeated back onto every original row**.

That means:
- same number of rows as original dataframe
- so use `transform`, not `agg`


In [5]:
df["dept_avg_salary"] = df.groupby("dept")["salary"].transform("mean")

df[["emp_id", "name", "dept", "salary", "dept_avg_salary"]]

,emp_id,name,dept,salary,dept_avg_salary
0,101,Alice,Data,120000,117500.000000
1,102,Bob,Data,95000,117500.000000
2,103,Charlie,Web,110000,104666.666667
3,104,David,Web,105000,104666.666667
4,105,Eva,Data,130000,117500.000000
5,106,Frank,ML,150000,148333.333333
6,107,Grace,ML,140000,148333.333333
7,108,Helen,Web,99000,104666.666667
8,109,Ian,Data,125000,117500.000000
9,110,Jane,ML,155000,148333.333333


### Explanation

For the `Data` department, the average salary is `117500`.  
So every `Data` employee gets `117500` in the new column.

Memory trick:
- `agg()` shrinks
- `transform()` broadcasts back


## 5. Task 4 — Compare employee salary to department average

### Goal

Create a new column:
- `salary_vs_dept_avg`

Logic:
- `"above"` if employee salary > department average
- `"below_or_equal"` otherwise

### Why this uses `np.where`
`np.where(condition, value_if_true, value_if_false)` is a fast way to build conditional columns.


In [6]:
df["salary_vs_dept_avg"] = np.where(
    df["salary"] > df["dept_avg_salary"],
    "above",
    "below_or_equal"
)

df[["emp_id", "name", "dept", "salary", "dept_avg_salary", "salary_vs_dept_avg"]]

,emp_id,name,dept,salary,dept_avg_salary,salary_vs_dept_avg
0,101,Alice,Data,120000,117500.000000,above
1,102,Bob,Data,95000,117500.000000,below_or_equal
2,103,Charlie,Web,110000,104666.666667,above
3,104,David,Web,105000,104666.666667,above
4,105,Eva,Data,130000,117500.000000,above
5,106,Frank,ML,150000,148333.333333,above
6,107,Grace,ML,140000,148333.333333,below_or_equal
7,108,Helen,Web,99000,104666.666667,below_or_equal
8,109,Ian,Data,125000,117500.000000,above
9,110,Jane,ML,155000,148333.333333,above


### Explanation

For Bob in `Data`:
- salary = 95000
- department average = 117500
- result = `below_or_equal`

For Eva in `Data`:
- salary = 130000
- department average = 117500
- result = `above`


## 6. Task 5 — Count high performers by department

### Goal

For each department, count employees where:
- `rating >= 4`

### Why this uses conditional aggregation
We group by department, then inside each group we count how many rows satisfy a condition.


In [7]:
high_performers_by_dept = df.groupby("dept").agg(
    high_performers=("rating", lambda x: (x >= 4).sum())
).reset_index()

high_performers_by_dept

,dept,high_performers
0,Data,3
1,ML,3
2,Web,1


### Explanation

Inside each department:
- `(x >= 4)` creates booleans
- `True` counts as `1`
- `False` counts as `0`
- `.sum()` gives total high performers

Example for `ML`, ratings are:
- 4, 5, 5

Condition `>= 4` becomes:
- True, True, True

Sum = `3`


## 7. Task 6 — Percentage of high performers by department

### Goal

For each department, compute:
- total employees
- high performers
- percentage of high performers

### Pattern
This is a two-step pattern:
1. aggregate counts
2. create a derived percentage column


In [8]:
high_perf_pct = df.groupby("dept").agg(
    total_employees=("emp_id", "count"),
    high_performers=("rating", lambda x: (x >= 4).sum())
).reset_index()

high_perf_pct["pct_high_performers"] = (
    high_perf_pct["high_performers"] / high_perf_pct["total_employees"] * 100
)

high_perf_pct

,dept,total_employees,high_performers,pct_high_performers
0,Data,4,3,75.000000
1,ML,3,3,100.000000
2,Web,3,1,33.333333


### Explanation

For `Web`:
- total employees = 3
- high performers = 1
- percentage = `1 / 3 * 100 = 33.33`

This kind of question is common in interviews:
- % premium users by country
- % passed students by class
- % fraud cases by region


## 8. Task 7 — Top paid employee in each department

### Goal

Return the full row of the highest salary employee per department.

There are two good interview-friendly ways to solve this:
1. `groupby + nlargest`
2. `sort_values + drop_duplicates`


### Method 1 — `groupby + nlargest`

In [9]:
top_paid_by_dept = df.groupby("dept", group_keys=False).apply(
    lambda g: g.nlargest(1, "salary")
)

top_paid_by_dept[["emp_id", "name", "dept", "salary"]]

/tmp/ipykernel_2824/390576734.py:1: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  top_paid_by_dept = df.groupby("dept", group_keys=False).apply(


,emp_id,name,dept,salary
4,105,Eva,Data,130000
9,110,Jane,ML,155000
2,103,Charlie,Web,110000


### Explanation

Within each department group:
- sort by salary descending
- keep the largest 1 row

This returns the full employee row, not just the salary.


### Method 2 — `sort_values + drop_duplicates`

In [10]:
top_paid_by_dept_2 = (
    df.sort_values(["dept", "salary"], ascending=[True, False])
      .drop_duplicates("dept")
)

top_paid_by_dept_2[["emp_id", "name", "dept", "salary"]]

,emp_id,name,dept,salary
4,105,Eva,Data,130000
9,110,Jane,ML,155000
2,103,Charlie,Web,110000


### Explanation

This works because:
1. sort each department so the highest salary comes first
2. drop duplicates by department
3. first row kept per department is the top paid one

This method is also very common in interviews.


## Final cheat sheet summary

### Use these patterns clearly

#### 1. Summary per group
```python
df.groupby("dept").agg(total_salary=("salary", "sum"))
```

#### 2. Summary by multiple keys
```python
df.groupby(["dept", "city"], as_index=False).agg(avg_salary=("salary", "mean"))
```

#### 3. Group metric back to each row
```python
df["dept_avg_salary"] = df.groupby("dept")["salary"].transform("mean")
```

#### 4. Condition using a group metric
```python
df["flag"] = np.where(df["salary"] > df["dept_avg_salary"], "above", "below")
```

#### 5. Conditional count inside groupby
```python
df.groupby("dept").agg(high_performers=("rating", lambda x: (x >= 4).sum()))
```

#### 6. Top row per group
```python
df.groupby("dept", group_keys=False).apply(lambda g: g.nlargest(1, "salary"))
```

### Interview memory line
- `agg` = one row per group  
- `transform` = same row count  
- `apply` = custom group logic  
- `np.where` = if/else column creation
